Add interface for lookup up of externally stored type descriptors

augusto2112 opened the pull request:
Currently, TypeRefBuilder knows how to parse type descriptors from reflection metadata. We aim to store the same information that lives in type descriptors externally, as an effort to support embedded Swift debugging. In order to take advantage of all of the existing type information parsing code that exists today in remote mirrors, add an interface for external lookup of type descriptors, and replace all the usages of type descriptors with their generic counterpart (this patch only adds support for builtin descriptors, but follow up patches will add support for other descriptor types).

augusto2112 commented:
@slavapestov this is my approach for looking up descriptors that live outside of reflection metadata. With this approach we can keep the entire type information building logic inside remote mirrors, and only implement in LLDB the logic to look up the descriptors.

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
https://github.com/apple/llvm-project/pull/7741

@swift-ci smoke test

augusto2112 committed:
Add interface for lookup up of externally stored type descriptors

Currently, TypeRefBuilder knows how to parse type descriptors from
reflection metadata. We aim to store the same information that lives
in type descriptors externally, as an effort to support embedded Swift
debugging. In order to take advantage of all of the existing type
information parsing code that exists today in remote mirrors, add an
interface for external lookup of type descriptors, and replace all the
usages of type descriptors with their generic counterpart (this patch
only adds support for builtin descriptors, but follow up patches will
add support for other descriptor types).

augusto2112 commented:
@swift-ci smoke test

adrian-prantl commented on include/swift/RemoteInspection/ExternalDescriptorFinder.h:
```diff
@@ -0,0 +1,56 @@
+
+enum class FieldDescriptorKind : uint16_t;
+
+/// A generic interface a builtin type descriptor.
```
missing word: "for"

adrian-prantl replied:
s/generic/abstract/

adrian-prantl commented on include/swift/RemoteInspection/ExternalDescriptorFinder.h:
```diff
@@ -0,0 +1,56 @@
+
+  virtual ~BuiltinTypeDescriptorInterface() {};
+
+  virtual llvm::StringRef readMangledTypeName() = 0;
```
`getMangledTypeName()`?
the "read" part is only true for the reflection metadata implementation, right?

adrian-prantl commented on include/swift/RemoteInspection/ExternalDescriptorFinder.h:
```diff
@@ -0,0 +1,56 @@
+  virtual llvm::StringRef readMangledTypeName() = 0;
+};
+
+// Interface for finding type descriptors that lives outside of reflection
```
///

adrian-prantl commented on include/swift/RemoteInspection/ExternalDescriptorFinder.h:
```diff
@@ -0,0 +1,56 @@
+
+// Interface for finding type descriptors that lives outside of reflection
+// metadata.
+struct ExternalDescriptorFinder {
```
The "external" part of the name seems unnecessary to me? What about TypeDescriptorLookup / TypeDescriptorLoader / TypeDescriptorFinder?

adrian-prantl commented on include/swift/RemoteInspection/ExternalDescriptorFinder.h:
```diff
@@ -0,0 +1,56 @@
+  virtual ~ExternalDescriptorFinder(){};
+
+  virtual std::unique_ptr<BuiltinTypeDescriptorInterface>
+  getBuiltinTypeInfoInterface(const TypeRef *TR) = 0;
```
1. You are not returning an interface, but an object, so this should be called (get|lookup|find)BuiltinTypeInfo().
2. Why can you know what kind of TypeDescriptor you need for a given TR? Is that unambigous based on the kind of TR?

augusto2112 replied:
Re 2: yes, it's unambiguous given the TR.

adrian-prantl commented on stdlib/public/RemoteInspection/TypeRefBuilder.cpp:
```diff
@@ -460,6 +461,37 @@ TypeRefBuilder::getBuiltinTypeInfo(const TypeRef *TR) {
+    if (auto descriptor = DF->getBuiltinTypeInfoInterface(TR))
+      return descriptor;
+
+  auto BTI = getBuiltinTypeInfo(TR);
```
Why isn't the second part of this function encapsulated in a second TypeDescriptorFinder that finds types in ReflectionMetadata?

augusto2112 replied:
Ok, I separated this into its own class. I have it on its own NFC commit so it might be easier to review it that way

adrian-prantl commented on include/swift/RemoteInspection/TypeLowering.h:
```diff
@@ -174,7 +175,7 @@ class BuiltinTypeInfo : public TypeInfo {
 public:
   explicit BuiltinTypeInfo(TypeRefBuilder &builder,
-                           RemoteRef<BuiltinTypeDescriptor> descriptor);
+                           BuiltinTypeDescriptorInterface *descriptor);
```
Why does this need to be a pointer? Could it at least be a reference?

augusto2112 committed:
[NFC] Extract descriptor finder from the rest of TypeRefBuilder

Separate the functionality that reads descriptors from reflection
information into its own class.

augusto2112 committed:
[NFC] clang-format TypeRefBuilder.h

augusto2112 committed:
[NFC] Rename DF -> RDF

augusto2112 commented:
@swift-ci smoke test

adrian-prantl commented on include/swift/RemoteInspection/DescriptorFinder.h:
```diff
@@ -0,0 +1,54 @@
+/// Interface for finding type descriptors. Implementors may provide descriptors
+/// that live inside or outside reflection metadata.
+struct DescriptorFinder {
+  virtual ~DescriptorFinder(){};
```
clang-format

augusto2112 replied:
That's the output of clang-format

adrian-prantl commented on include/swift/RemoteInspection/DescriptorFinder.h:
```diff
@@ -0,0 +1,54 @@
+struct DescriptorFinder {
+  virtual ~DescriptorFinder(){};
+
+  virtual std::unique_ptr<BuiltinTypeDescriptorInterface>
```
Performance-wise, you could design the base class to already have all the data members needed by all implementations, maybe with an opaque element for implementation-specific data, then you could pass a stack-allocated object through this interface and don't need to use unique_ptr here.

augusto2112 replied:
Personally I don't think the performance benefits would be worth the extra complexity. There's already indirection going on today when using `RemoteRef`, and there will be a pretty big difference in size between the DWARF version and the reflection version, so accommodating both on the base class would look a bit ugly imo.

adrian-prantl replied:
If this is only used in the debugger and similar clients, I think that's reasonable. If it's used in-process then I'm not sure.

augusto2112 replied:
This is only used for out of process inspection.

adrian-prantl commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -38,7 +39,8 @@ namespace reflection {
 
-template <typename Runtime> class ReflectionContext;
+template <typename Runtime>
+class ReflectionContext;
```
Can you commit a separate patch that just runs clang-format on this file? It would reduce the size of the diff considerably.

augusto2112 replied:
There's already a commit that just clang formats this file

augusto2112 replied:
There are three NFC commits and one commit with the actual changes

adrian-prantl replied:
You're right of course!

adrian-prantl commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -486,6 +455,368 @@ class TypeRefBuilder {
     }
   };
 
+  struct ReflectionTypeDescriptorFinder
```
Doxygen comment?

adrian-prantl commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -486,6 +455,368 @@ class TypeRefBuilder {
+  struct ReflectionTypeDescriptorFinder
+      : public swift::reflection::DescriptorFinder {
+    ReflectionTypeDescriptorFinder(TypeRefBuilder &Builder,
+                                   remote::ExternalTypeRefCache *externalCache)
```
comment what externalCache is supposed to be?

augusto2112 replied:
That's not related to this patch, but added a comment in the ivar explaining what that is.

adrian-prantl commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -486,6 +455,368 @@ class TypeRefBuilder {
+    template <template <typename Runtime> class ObjCInteropKind,
+              unsigned PointerSize>
+    AssociatedTypeCollectionResult
+    collectAssociatedTypes(llvm::Optional<std::string> forMangledTypeName) {
```
does this have to be in the .h file, or could this be moved into a .cpp file?

adrian-prantl replied:
(and maybe the answer is yes, because of the template args)

augusto2112 replied:
Yes, it needs to be in the .h due to the template args

adrian-prantl commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -518,40 +849,34 @@ class TypeRefBuilder {
 
-  BuiltProtocolDecl
-  createProtocolDecl(Node *node) {
+  BuiltProtocolDecl createProtocolDecl(Node *node) {
```
same comment about running clang-format in a preparatory commit

adrian-prantl reviewed (commented):
In general I think this looks promising, but the diff is so large due to the formatting changes that it's a bit difficult to review.

augusto2112 commented:
@swift-ci smoke test

adrian-prantl commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -486,6 +465,363 @@ class TypeRefBuilder {
+        NormalizedReflectionNameCache;
+
+    /// Cache for built-in type descriptor lookups.
+    std::unordered_map<std::string /* normalized name */,
```
Not for this patch, but since we have LLVM here, this should probably be a StringMap for better efficiency. Same for all the other caches.

adrian-prantl commented on stdlib/public/RemoteInspection/TypeRefBuilder.cpp:
```diff
@@ -471,6 +471,42 @@ TypeRefBuilder::ReflectionTypeDescriptorFinder::getBuiltinTypeInfo(
   return nullptr;
 }
 
+class BuiltinTypeDescriptorImpl : public BuiltinTypeDescriptorBase {
```
Doxygen comment?

adrian-prantl commented on stdlib/public/RemoteInspection/TypeRefBuilder.cpp:
```diff
@@ -471,6 +471,42 @@ TypeRefBuilder::ReflectionTypeDescriptorFinder::getBuiltinTypeInfo(
+
+std::unique_ptr<BuiltinTypeDescriptorBase>
+TypeRefBuilder::getBuiltinTypeDescriptor(const TypeRef *TR) {
+  if (DF) 
```
Since you are going to write a bunch these, maybe you should have a SmallVector<DescriptorFInderBase&, 2> DFs and say 

```
for (DF : DFs) 
  if (auto D = DF->getBuiltinTypeDescriptor(TR))
      return D;
```
?

adrian-prantl replied:
Maybe it's overkill because you'll have exactly 2...

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
@swift-ci smoke test
